### DISCUSSION TITLE
Plugin-client-redirects: How does `createRedirects` work?

### DISCUSSION THREAD

**Discussion by thomasheartman** [facebook/docusaurus#8337] [Q&A] [answered]
Hey! I'm using plugin-client-redirects and the `createRedirects` function, I can't seem to make the redirects work or be able to competently figure out why they're not working. Could any of you help me understand why and/or what I'm doing wrong?

I'm aware that redirects don't work in dev, so I'm building the docs with `docusaurus build` and serving with `docusaurus serve` to try it out. Redirects declared within redirects work, but the `createRedirects` property doesn't seem to have any effect. This is with version 2.2.0.

What I'm trying to do is effectively redirect /addons/doc to /reference/addons/doc with something that looks like this:
```js
                createRedirects(existingPath) {
                    if (existingPath.startsWith('/addons')) {
                        return '/reference' + existingPath;
                    }

                    return undefined;
                },
```

I've also tried using `includes`, logging `existingPath` using `console.log`, and just willy-nilly rewriting anything to see if anything happens, but nothing seems to take effect.

Also, if there's a better way to achieve that redirect, then please let me know 😄

**Comment by johnnyreilly** [comment, marked as answer]
Heya,

I used client redirects on my site until fairly recently, you can see how I was using them here:

https://blog.johnnyreilly.com/definitive-guide-to-migrating-from-blogger-to-docusaurus#redirecting-from-blogger-urls-to-docusaurus-urls

Hopefully this provides some kind of guide as to how to use them?

**Reply by thomasheartman** [reply]
Hey! Thanks for the response and for the link 😄 It _did_ provide some guidance, but I'm still stuck:

You made me realize that I'd misread the docs earlier (or skimmed over the code comments): the input argument is an existing URL path that docusaurus has created. And what you return is the `fromUrl`.

However, it still doesn't seem to work for me. There's no redirection happening, it seems. I also noticed your comment about the `console.log` being emitted at build time (makes sense when the function works this way), but I'm not seeing anything. It could be that it's writing to the console and then immediately overwriting it with a different message though.

Do you know if there's a good way to debug what's going on?

**Reply by johnnyreilly** [reply]
`console.log` at build time is probably the only debugging tool at your disposal here I'm afraid. If I was you I'd make that your first statement in the function and see what results

**Reply by thomasheartman** [reply]
Good suggestion (that I should honestly have thought of myself 😅 Sadly, even if it's the first line of code in that function: nothing seems to be happening 🤷🏼  Nothing is logged as far as I can tell.

**Reply by johnnyreilly** [reply]
I suspect you may have some issue with your docusaurus site / config.  Might be worth spinning up a new instance and experimenting there.

**Reply by thomasheartman** [reply]
Yeah, I think you may be right. I ended up just writing the redirects out manually (there were only five of them, anyway), so I worked around it. But later on I realized that we already had a `createRedirects` function already declared (a bit later), so I'm guessing that was the issue. And rather than failing to build because of the duplicate property, it just silently picked the one that was declared later 🙈 

I wish docusaurus would've let me know, but it's probably not common enough for it to be a prioritized thing to work on. 

Anyway, thanks for all the input; I'll mark it as answered 🙌🏼

**Reply by slorber** [reply]
Hey

If you want to redirect from `/addons/doc` to `/reference/addons/doc`

You can debug this by looking at a static file in  `/addons/doc/index.html` (or `/addons/doc.html`) 

=> it should contain a some basic JS code doing `window.location = "/reference/addons/doc"`

If it's not the case, please share your existing client redirect config.

Note the client redirect plugin will not attempt to "override" an existing HTML file, so if you already have something on route `/addons/doc`, afaik it won't be overridden by a redirect and will be silently ignored to prevent mistakes. 

Now we can add options to make this behavior fill-fast in case an override is detected, but it would be nice to create a repro first to understand what exactly is happening in your case

**Reply by thomasheartman** [reply]
Oh, checking the generated output for an actual redirect file is clever! 🧠 

> Now we can add options to make this behavior fill-fast in case an override is detected, but it would be nice to create a repro first to understand what exactly is happening in your case

Yeah, so what was happening in my case was that I was redeclaring a property that already existed on the object. It went a little something like this:

```js
{
  createRedirects(existingPath) {
      if (existingPath.startsWith('/addons')) {
          return '/reference' + existingPath;
      }

      return undefined;
  },
  redirects: [
    // long list of manual redirects
  ],
  createRedirects(existingPath) {
    // pre-existing function that I didn't see
  }
}
```

When I found out, I had already worked around my issue, so I didn't try and declare the redirects in the pre-existing function, but my guess is that that would've worked just fine.

As far as I can tell, that was the root cause of this issue. JS doesn't really care if you declare the same property multiple times, so it makes sense that it gets overridden. I don't know if there is anything docusaurus _can_ do to detect this, but it would be nice to get a warning in the build process for it. That said, I can imagine that it'll require more effort than it may be worth 🤷🏼

**Reply by slorber** [reply]
I see thanks.

I guess it's possibly the role of ESLint to give you an error here, and you'll have to configure it yourself as we don't impose the usage of ESLint alongside a Docusaurus site. (the Facebook site template has ESLint/Prettier preconfigured FYI)

**Reply by thomasheartman** [reply]
Aye, I'd agree that ESLint should provide an error here. I don't know why it didn't (or maybe I just didn't see it?). However, that'll only work if the user uses ESLint, which, while probably the norm, is probably not always the case. Again, it'd be super nice if docusaurus would drop a warning about this during the build (or if you had a setting where you could configure what to do "on duplicate config properties", such as `"ignore" | "warn" | "error" | "throw"`).  But again, with how JS happily accepts it, it's probably not very easy. 

But I don't know how often this sort of error crops up, so I don't know how much real-world value it'd provide.

**Reply by slorber** [reply]
Docusaurus does not see any duplicate property at runtime. Only a static analysis tool will tell you, and that still seems out of the scope of Docusaurus.
